src/stream/src/executor/source/source_backfill_executor.rs

```diff
@@ -665,11 +648,25 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
                                         .await?;
                                     }
 
+                                    if !state_table_initialized {
+                                        // This is for self.backfill_finished() to be safe: wait until this actor can read all actors' written data.
+                                        // We wait for 2nd epoch
+                                        let epoch = barrier_epoch.prev;
```

Comment:
The condition for `backfill_finished()` is that all loaded states are at `Finished` state. Is the reason for potential correctness issue that, when we fail to load the initial states written by other actors, we may get something like an empty list and then in the empty list, and the all-finished condition can be satisfied?

Reply:
I'm thinking that shall we let all source parallelisms be aware of globally all the source splits instead of only the source splits they own? This global source split list can also be updated and maintained through barrier mutation. And then here instead of iter to load the all states written to storage, we do point-get one by one to explicitly see if all source splits exist and have finished.

Reply:
> when we fail to load the initial states written by other actors, we may get something like an empty list and then in the empty list, and the all-finished condition can be satisfied?

exactly

> This global source split list can also be updated and maintained through barrier mutation

I'm a little confused how to implement this :eyes:

Reply:
> > This global source split list can also be updated and maintained through barrier mutation
> 
> I'm a little confused how to implement this 👀

Currently we create actors along with barrier injection, so we can include the list of all source splits at the moment in the create actor information, and the source split change mutation will include all the changes, and then we can update source splits list.

Reply:
Oh, you are quite right. I got it

Reply:
> and the source split change mutation will include all the changes, and then we can update source splits list.

you mean carrying the fragment_id along with the assignments? I don't get why each actor need to maintain the whole upstream split list.